Add dot output to jlm-opt, using the GraphWriter API

haved opened the pull request:
Adds the command line option `--output-format=dot`, which causes the final RVSDG to be dumped as the GraphViz dot format.

One non-standard thing about the output is that each region becomes a separate graph, but with globally unique ids. This is done to make it easier to focus on parts of the RVSDG at a time, but it requires a renderer that expects multiple graphs in a single file.

I have made a webapp that takes concatenated graphs as input, and creates a tree based on which graphs are referenced as subgraphs in other graphs. Selecting a node, edge or port will show all its properties, and highlight other nodes/edges they reference. The viewer should also work with any other GraphViz you may have.

See the webapp [here](https://dot-tree-viz.vercel.app/).
Repo for webapp [here](https://github.com/haved/dot-tree-viz).

The webapp needs more polish to make it pleasant to work with complex graphs (resizing views, rendering subgraphs inside graphs, rendering \n properly).

<details>
<summary>Example output (long)</summary>

```
digraph graph0 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  label="Type graph"
  
  node0 [label=ptr tooltip="5e0edd496b28" ];
  node1 [label=iostate tooltip="5e0edd496b38" ];
  node2 [label=mem tooltip="5e0edd496b40" ];
  node3 [label=bit32 tooltip="5e0edd496e80" ];
  node4 [label="ctl(2)" tooltip="5e0edd499200" ];
  node5 [label="ctl(3)" tooltip="5e0edd499210" ];
  node6 [label=bit1 tooltip="5e0edd496c90" ];
}
digraph graph1 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  label="RVSDG root graph"
  tooltip="5e0f07188b80" 
  node7 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">LAMBDA[lzma_decode]</TD></TR>
                        <TR><TD>
                                <TABLE BORDER="0" CELLSPACING="4" CELLPADDING="2"><TR>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph2">graph2</TD>
                                </TR></TABLE>
                        </TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o0" type="node0" BGCOLOR="#FFFFFF" tooltip="5e0f071a73f0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718e890" ];
  {
    rank=sink;
    r13 [label=r13 tooltip="5e0f0719eec0" ];
  }
  node7:o0:s -> r13[id=edge48 ];
}
digraph graph2 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f07184ac0" 
  {
    rank=source;
    a0 [label=a0 type=node1 tooltip="5e0f071a0b00" ];
    a1 [label=a1 type=node2 tooltip="5e0f071a0c80" ];
    a0 -> a1[style=invis];
  }
  node8 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(2)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o1" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a3110"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a70c0" ];
  node9 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">undef</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o2" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a21b0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a20e0" ];
  node10 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o3" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a1600"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718f7d0" ];
  node11 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o4" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a0f10"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718e420" ];
  node12 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">undef</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o5" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a0370"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718e9c0" ];
  node13 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i0" BGCOLOR="#FFFFFF" tooltip="5e0f0719cb00"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">ALLOCA[bit32]</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o6" type="node0" BGCOLOR="#FFFFFF" tooltip="5e0f071a10d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o7" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a1160"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718f710" ];
  node14 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i1" BGCOLOR="#FFFFFF" tooltip="5e0f0719c8e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i2" BGCOLOR="#FFFFFF" tooltip="5e0f0719c520"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">MemoryStateMerge</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o8" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a1550"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718e760" ];
  node15 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i3" BGCOLOR="#FFFFFF" tooltip="5e0f071a1710"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i4" BGCOLOR="#FFFFFF" tooltip="5e0f071a17e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i5" BGCOLOR="#FFFFFF" tooltip="5e0f071a18d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Store</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o9" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a1980"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f07192640" ];
  node16 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i6" BGCOLOR="#FFFFFF" tooltip="5e0f071a1a80"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i7" BGCOLOR="#FFFFFF" tooltip="5e0f071a1ae0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Load</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o10" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a1bd0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o11" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a1c40"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071926d0" ];
  node17 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i8" BGCOLOR="#FFFFFF" tooltip="5e0f071a1f20"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">MATCH[1 -&gt; 1, 0 -&gt; 0, 2]</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o12" type="node5" BGCOLOR="#FFFFFF" tooltip="5e0f071a2010"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a1da0" ];
  node18 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i9" BGCOLOR="#FFFFFF" tooltip="5e0f071a03e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">GAMMA</TD></TR>
                        <TR><TD>
                                <TABLE BORDER="0" CELLSPACING="4" CELLPADDING="2"><TR>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph3">graph3</TD>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph4">graph4</TD>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph5">graph5</TD>
                                </TR></TABLE>
                        </TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o13" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a2bf0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f0718e590" ];
  node22 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i10" BGCOLOR="#FFFFFF" tooltip="5e0f0719a8e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i11" BGCOLOR="#FFFFFF" tooltip="5e0f0718b2d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i12" BGCOLOR="#FFFFFF" tooltip="5e0f071a3180"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">GAMMA</TD></TR>
                        <TR><TD>
                                <TABLE BORDER="0" CELLSPACING="4" CELLPADDING="2"><TR>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph6">graph6</TD>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph7">graph7</TD>
                                </TR></TABLE>
                        </TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o17" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a3560"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071924c0" ];
  {
    rank=sink;
    r10 [label=r10 tooltip="5e0f071a72d0" ];
    r11 [label=r11 tooltip="5e0f071a7320" ];
    r10 -> r11[style=invis];
    r12 [label=r12 tooltip="5e0f071a7370" ];
    r11 -> r12[style=invis];
  }
  node11:o4:s -> node13:i0:n[id=edge32 ];
  node13:o7:s -> node14:i1:n[id=edge33 color="#FF0000" ];
  a1 -> node14:i2:n[id=edge34 color="#FF0000" ];
  node13:o6:s -> node15:i3:n[id=edge35 ];
  node10:o3:s -> node15:i4:n[id=edge36 ];
  node14:o8:s -> node15:i5:n[id=edge37 color="#FF0000" ];
  node13:o6:s -> node16:i6:n[id=edge38 ];
  node15:o9:s -> node16:i7:n[id=edge39 color="#FF0000" ];
  node16:o10:s -> node17:i8:n[id=edge40 ];
  node17:o12:s -> node18:i9:n[id=edge41 ];
  node18:o13:s -> node22:i10:n[id=edge42 ];
  node13:o6:s -> node22:i11:n[id=edge43 ];
  node16:o11:s -> node22:i12:n[id=edge44 color="#FF0000" ];
  node8:o1:s -> r10[id=edge45 ];
  a0 -> r11[id=edge46 color="#00FF00" ];
  node22:o17:s -> r12[id=edge47 color="#FF0000" ];
}
digraph graph3 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a2260" 
  node19 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">CTL(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o14" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a26a0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a25c0" ];
  {
    rank=sink;
    r0 [label=r0 tooltip="5e0f071a2c80" ];
  }
  node19:o14:s -> r0[id=edge0 ];
}
digraph graph4 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a2320" 
  node20 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">CTL(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o15" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a2770"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a28a0" ];
  {
    rank=sink;
    r1 [label=r1 tooltip="5e0f071a2cd0" ];
  }
  node20:o15:s -> r1[id=edge1 ];
}
digraph graph5 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a23e0" 
  node21 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">CTL(0)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o16" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a29c0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a2a50" ];
  {
    rank=sink;
    r2 [label=r2 tooltip="5e0f071a2dd0" ];
  }
  node21:o16:s -> r2[id=edge2 ];
}
digraph graph6 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a2ef0" 
  {
    rank=source;
    a2 [label=a2 type=node0 tooltip="5e0f0719c1c0" ];
    a3 [label=a3 type=node2 tooltip="5e0f0719aac0" ];
    a2 -> a3[style=invis];
  }
  {
    rank=sink;
    r3 [label=r3 tooltip="5e0f071a35f0" ];
  }
  a3 -> r3[id=edge3 color="#FF0000" ];
}
digraph graph7 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a2fb0" 
  {
    rank=source;
    a4 [label=a4 type=node0 tooltip="5e0f0719c2b0" ];
    a5 [label=a5 type=node2 tooltip="5e0f0719ae60" ];
    a4 -> a5[style=invis];
  }
  node23 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i13" BGCOLOR="#FFFFFF" tooltip="5e0f071a3810"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i14" BGCOLOR="#FFFFFF" tooltip="5e0f071a3bf0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">THETA</TD></TR>
                        <TR><TD>
                                <TABLE BORDER="0" CELLSPACING="4" CELLPADDING="2"><TR>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph8">graph8</TD>
                                </TR></TABLE>
                        </TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o18" type="node0" BGCOLOR="#FFFFFF" tooltip="5e0f071a3920"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o19" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a3d00"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a33b0" ];
  {
    rank=sink;
    r9 [label=r9 tooltip="5e0f071a7070" ];
  }
  a4 -> node23:i13:n[id=edge29 ];
  a5 -> node23:i14:n[id=edge30 color="#FF0000" ];
  node23:o19:s -> r9[id=edge31 color="#FF0000" ];
}
digraph graph8 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a3480" 
  {
    rank=source;
    a6 [label=a6 type=node0 tooltip="5e0f07195020" ];
    a7 [label=a7 type=node2 tooltip="5e0f0719ab40" ];
    a6 -> a7[style=invis];
  }
  node24 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i15" BGCOLOR="#FFFFFF" tooltip="5e0f071a4020"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i16" BGCOLOR="#FFFFFF" tooltip="5e0f071a4080"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Load</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o20" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a4100"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o21" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a4170"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a3ee0" ];
  node25 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(5)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o22" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a67b0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a6690" ];
  node26 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(10)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o23" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a4360"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a4240" ];
  node27 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i17" BGCOLOR="#FFFFFF" tooltip="5e0f071a4670"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i18" BGCOLOR="#FFFFFF" tooltip="5e0f071a4760"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BitSlt32</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o24" type="node6" BGCOLOR="#FFFFFF" tooltip="5e0f071a4850"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a4550" ];
  node28 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i19" BGCOLOR="#FFFFFF" tooltip="5e0f071a4a20"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">MATCH[1 -&gt; 1, 0]</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o25" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a4b10"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a48e0" ];
  node29 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i20" BGCOLOR="#FFFFFF" tooltip="5e0f071a4df0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i21" BGCOLOR="#FFFFFF" tooltip="5e0f071a4ee0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i22" BGCOLOR="#FFFFFF" tooltip="5e0f071a5030"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">GAMMA</TD></TR>
                        <TR><TD>
                                <TABLE BORDER="0" CELLSPACING="4" CELLPADDING="2"><TR>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph9">graph9</TD>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph10">graph10</TD>
                                </TR></TABLE>
                        </TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o26" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a6320"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a4be0" ];
  node34 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i30" BGCOLOR="#FFFFFF" tooltip="5e0f071a5150"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i31" BGCOLOR="#FFFFFF" tooltip="5e0f071a65a0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Load</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o32" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a52e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o33" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a65e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a64c0" ];
  node35 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i32" BGCOLOR="#FFFFFF" tooltip="5e0f071a69a0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i33" BGCOLOR="#FFFFFF" tooltip="5e0f071a6a90"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BitSlt32</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o34" type="node6" BGCOLOR="#FFFFFF" tooltip="5e0f071a6b80"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a6880" ];
  node36 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i34" BGCOLOR="#FFFFFF" tooltip="5e0f071a6dd0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">MATCH[1 -&gt; 1, 0]</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o35" type="node4" BGCOLOR="#FFFFFF" tooltip="5e0f071a6ec0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a6c10" ];
  {
    rank=sink;
    r6 [label=r6 tooltip="5e0f071a3710" ];
    r7 [label=r7 tooltip="5e0f071a39f0" ];
    r6 -> r7[style=invis];
    r8 [label=r8 tooltip="5e0f071a3d90" ];
    r7 -> r8[style=invis];
  }
  a6 -> node24:i15:n[id=edge13 ];
  a7 -> node24:i16:n[id=edge14 color="#FF0000" ];
  node24:o20:s -> node27:i17:n[id=edge15 ];
  node26:o23:s -> node27:i18:n[id=edge16 ];
  node27:o24:s -> node28:i19:n[id=edge17 ];
  node28:o25:s -> node29:i20:n[id=edge18 ];
  a6 -> node29:i21:n[id=edge19 ];
  node24:o21:s -> node29:i22:n[id=edge20 color="#FF0000" ];
  a6 -> node34:i30:n[id=edge21 ];
  node29:o26:s -> node34:i31:n[id=edge22 color="#FF0000" ];
  node34:o32:s -> node35:i32:n[id=edge23 ];
  node25:o22:s -> node35:i33:n[id=edge24 ];
  node35:o34:s -> node36:i34:n[id=edge25 ];
  node36:o35:s -> r6[id=edge26 ];
  a6 -> r7[id=edge27 ];
  node34:o33:s -> r8[id=edge28 color="#FF0000" ];
}
digraph graph9 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a4c90" 
  {
    rank=source;
    a8 [label=a8 type=node0 tooltip="5e0f0719c140" ];
    a9 [label=a9 type=node2 tooltip="5e0f0719b190" ];
    a8 -> a9[style=invis];
  }
  {
    rank=sink;
    r4 [label=r4 tooltip="5e0f071a63b0" ];
  }
  a9 -> r4[id=edge4 color="#FF0000" ];
}
digraph graph10 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="5e0f071a4d30" 
  {
    rank=source;
    a10 [label=a10 type=node0 tooltip="5e0f0719af20" ];
    a11 [label=a11 type=node2 tooltip="5e0f0719c4a0" ];
    a10 -> a11[style=invis];
  }
  node30 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i23" BGCOLOR="#FFFFFF" tooltip="5e0f071a54c0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i24" BGCOLOR="#FFFFFF" tooltip="5e0f071a5590"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Load</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o27" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a5680"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o28" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a56f0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a53b0" ];
  node31 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o29" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a58e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a57c0" ];
  node32 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i25" BGCOLOR="#FFFFFF" tooltip="5e0f071a5c80"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i26" BGCOLOR="#FFFFFF" tooltip="5e0f071a5d70"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BitAdd32</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o30" type="node3" BGCOLOR="#FFFFFF" tooltip="5e0f071a5e60"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a5b60" ];
  node33 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i27" BGCOLOR="#FFFFFF" tooltip="5e0f071a6040"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i28" BGCOLOR="#FFFFFF" tooltip="5e0f071a60a0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i29" BGCOLOR="#FFFFFF" tooltip="5e0f071a6190"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Store</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o31" type="node2" BGCOLOR="#FFFFFF" tooltip="5e0f071a6270"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="5e0f071a5f20" ];
  {
    rank=sink;
    r5 [label=r5 tooltip="5e0f071a6400" ];
  }
  a10 -> node30:i23:n[id=edge5 ];
  a11 -> node30:i24:n[id=edge6 color="#FF0000" ];
  node30:o27:s -> node32:i25:n[id=edge7 ];
  node31:o29:s -> node32:i26:n[id=edge8 ];
  a10 -> node33:i27:n[id=edge9 ];
  node32:o30:s -> node33:i28:n[id=edge10 ];
  node30:o28:s -> node33:i29:n[id=edge11 color="#FF0000" ];
  node33:o31:s -> r5[id=edge12 color="#FF0000" ];
}

```
</details>

haved commented:
I notice I misspelled my email address in the git config, see the last two commits. That "user" should disappear when this PR is squashed.

haved commented on jlm/tooling/CommandLine.cpp:
```diff
@@ -253,11 +253,9 @@ const std::unordered_map<JlmOptCommandLineOptions::OutputFormat, std::string_vie
-    { OutputFormat::Mlir, "mlir" },
-    { OutputFormat::Tree, "tree" },
-    { OutputFormat::Xml, "xml" }
+    { OutputFormat::Ascii, "ascii" }, { OutputFormat::Dot, "dot" },
```
This appears to be what `clang-format` prefers, at least with our current config.

phate commented on jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp:
```diff
@@ -0,0 +1,147 @@
+    auto & elementTypeNode = GetOrCreateTypeGraphNode(vectorType->type(), typeGraph);
+    typeGraph.CreateDirectedEdge(elementTypeNode, node);
+  }
+  // TODO: Function types are also aggregate types (arguments and results)
```
Could we add function types here as well (and other types we know are missing), and end the cascade with a:

```
else
{
  JLM_UNREACHABLE("Unhandled type");
}
```

phate commented on jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp:
```diff
@@ -0,0 +1,147 @@
+{
+  graph.SetProgramObject(region);
+
+  // Connects an input port in the GraphWriter graph to an input in the RVSDG.
+  // Also adds an edge from the input's origin to the input port in the graph.
+  auto AttachInput = [&](util::Port & inputPort, const rvsdg::input & rvsdgInput)
+  {
+    inputPort.SetProgramObject(rvsdgInput);
+
+    if (auto originPort =
+            reinterpret_cast<util::Port *>(graph.GetElementFromProgramObject(*rvsdgInput.origin())))
+    {
+      auto & edge = graph.CreateDirectedEdge(*originPort, inputPort);
+      if (rvsdg::is<MemoryStateType>(rvsdgInput.type()))
+        edge.SetAttribute("color", util::Colors::Red);
+      if (rvsdg::is<iostatetype>(rvsdgInput.type()))
+        edge.SetAttribute("color", util::Colors::Green);
+    }
+  };
+
+  // Connects an output port in the GraphWriter graph to an output in the RVSDG
+  // Also adds a type attribute referencing a node in the typeGraph, if one exists.
+  auto AttachOutput = [&](util::Port & outputPort, const rvsdg::output & rvsdgOutput)
+  {
+    outputPort.SetProgramObject(rvsdgOutput);
+    if (typeGraph)
+      outputPort.SetAttributeGraphElement(
+          "type",
+          GetOrCreateTypeGraphNode(rvsdgOutput.type(), *typeGraph));
+  };
```
Could we move them out as static functions to aid readability?

phate commented on jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp:
```diff
@@ -0,0 +1,147 @@
+      {
+        auto & subGraph = node.CreateSubgraph();
+        CreateGraphNodes(subGraph, *structuralNode->subregion(i), typeGraph);
+        // FIXME: Attach the arguments and results from the subgraph to node's inputs and outputs
+        // Requires special logic for each type of structural node
```
What would it take to fix this?

haved replied:
Honestly it might be more straightforward than i initially thought, as structural results and arguments reference their structural inputs and outputs when applicable

phate commented on jlm/llvm/Makefile.sub:
```diff
@@ -5,6 +5,7 @@ libllvm_SOURCES = \
     jlm/llvm/backend/jlm2llvm/instruction.cpp \
     jlm/llvm/backend/jlm2llvm/jlm2llvm.cpp \
     jlm/llvm/backend/jlm2llvm/type.cpp \
+    jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp \
```
Let's just call it `dot`: `jlm/llvm/backend/dot`. I know the other ones are named differently right now, but I want to move away from this.

phate commented on jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp:
```diff
@@ -0,0 +1,147 @@
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/rvsdg/type.hpp>
+
+namespace jlm::llvm::rvsdg2dot
```
The same with the namespace. I think just `dot` is sufficient.

phate commented on jlm/llvm/backend/rvsdg2dot/rvsdg2dot.hpp:
```diff
@@ -0,0 +1,29 @@
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_LLVM_BACKEND_RVSDG2DOT_RVSDG2DOT_HPP
```
The same about the file names.

phate reviewed (changes requested):
This looks great. I am really exited about this. However, could we also get some simple unit tests on the backend? Maybe one for the normal graphs and one for the type graphs?

haved commented:
@phate I changed the path of the file to `jlm/llvm/backend/dot/DotWriter.[ch]pp`, let me know if you wanted something different.

In order to add some tests, I modified the API of the GraphWriter a bit, fixing some things I regretted from the original API.

phate commented on jlm/llvm/backend/dot/DotWriter.cpp:
```diff
@@ -0,0 +1,206 @@
+  node.SetLabel(type.debug_string());
+
+  // Some types get special handling, such as adding incoming edges from aggregate types
+  if (jlm::rvsdg::is<rvsdg::statetype>(type) || jlm::rvsdg::is<rvsdg::bittype>(type)
+      || jlm::rvsdg::is<PointerType>(type) || jlm::rvsdg::is<fptype>(type)
+      || jlm::rvsdg::is<varargtype>(type))
```
More like FYI, but it should not be necessary to prefix everything with `jlm::`. `rvsdg::` should be sufficient.

phate replied:
If I am mistaken here, then it might be beneficial to use `using namespace jlm::rvsdg`.

haved replied:
You are right. For some reason CLion did not gray out the `jlm::` part.

phate commented on jlm/llvm/backend/dot/DotWriter.cpp:
```diff
@@ -0,0 +1,206 @@
+  }
+
+  // Create a node for each node in the region
+  rvsdg::topdown_traverser traverser(&region);
+  for (const auto rvsdgNode : traverser)
```
Does the order of traversal make a difference here? Could you not just use `(const auto rvsdgNode : region.nodes)`?

haved replied:
It does make a difference, as inputs need to add edges to the output representing the origin, which has to be created first. Will make it clearer

phate commented on tests/jlm/llvm/backend/dot/DotWriterTests.cpp:
```diff
@@ -0,0 +1,105 @@
+  dot::WriteGraphs(writer, *gammaTest.graph().root(), true);
+
+  writer.Finalize();
+  // writer.OutputAllGraphs(std::cout, GraphOutputFormat::Dot);
```
Let the line commented in, otherwise it will get stale. There is no problem in printing something in a test.

phate commented on tests/jlm/llvm/backend/dot/DotWriterTests.cpp:
```diff
@@ -0,0 +1,105 @@
+  GraphWriter writer;
+  dot::WriteGraphs(writer, *gammaTest.graph().root(), false);
+
+  // writer.OutputAllGraphs(std::cout, GraphOutputFormat::Dot);
```
Let the line commented in, otherwise it will get stale. There is no problem in printing something in a test.

haved commented:
@phate I implemented your feedback, please review again
